packages/core/assistant-testing/src/functions/discord/fetch-messages.ts

```diff
@@ -4,133 +4,195 @@
+      const { threads: guildThreads } = yield* rest.getActiveGuildThreads(serverId);
+      channels.push(...guildThreads);
       if (channelId) {
-        channelIds = [channelId];
-      } else if (serverId) {
-        const channels = yield* rest.listGuildChannels(serverId);
-        const { threads } = yield* rest.getActiveGuildThreads(serverId);
-        const allChannels = [...channels, ...threads];
-        log('allChannels', {
-          channels: allChannels.map((channel) => ({ id: channel.id, name: 'name' in channel && channel.name })),
-        });
-        channelIds = allChannels.map((channel) => channel.id);
-      } else {
-        throw new Error('serverId or channelId is required');
+        channels = channels.filter((channel) => channel.id === channelId);
+      }
+      if (channels.length === 0) {
+        throw new Error('no channels found');
+      }
+      for (const channel of channels) {
+        console.log(channel.id, 'name' in channel ? channel.name : undefined);
       }
 
-      yield* TracingService.emitStatus({ message: `Will fetch from channels: ${channelIds.length}` });
-
-      const allMessages: DataType.Message[] = [];
-      for (const channelId of channelIds) {
-        let lastMessage: Option.Option<DataType.Message> = Option.none();
-        while (true) {
-          const { id: lastId = undefined } = pipe(
-            lastMessage,
-            Option.map(Obj.getKeys('discord.com')),
-            Option.flatMap(Option.fromIterable),
-            Option.getOrElse(() => ({ id: undefined })),
-          );
-
-          const options = {
-            after: !lastId ? `${generateSnowflake(after)}` : lastId,
-            limit: pageSize,
-          };
-          log('fetching messages', {
-            lastId,
-            afterSnowflake: options.after,
-            after: parseSnowflake(options.after),
-            limit: options.limit,
-          });
-          const messages = yield* rest.listMessages(channelId, options).pipe(
-            Effect.map(Array.map(makeMessage)),
-            Effect.map(Array.reverse),
-            Effect.catchTag('ErrorResponse', (err) =>
-              err.cause.code === 50001 ? Effect.succeed([]) : Effect.fail(err),
-            ),
-          );
-          if (messages.length > 0) {
-            lastMessage = Option.fromNullable(messages.at(-1));
-            allMessages.push(...messages);
-          } else {
-            break;
-          }
-          yield* TracingService.emitStatus({ message: `Fetched messages: ${allMessages.length}` });
-          if (allMessages.length >= limit) {
-            break;
+      yield* TracingService.emitStatus({ message: `Will fetch from channels: ${channels.length}` });
+
+      const threads = yield* Effect.forEach(
+        channels,
+        Effect.fnUntraced(function* (channel) {
+          const allMessages: DataType.Message[] = [];
+
+          let lastMessage: Option.Option<DataType.Message> = Option.none();
```

Comment:
given this always starts from none I assume this sync is one shot and can't be run on a schedule without creating duplicate data?

Reply:
you can specify `after` timestamp manually in the call parameters, but this doesn't sync with a particular queue.